Improved itermonomials

ctsiagkalis opened the pull request:
#### Brief description of what is fixed or changed

itermonomials now computes only the necessary monomials and returns a generator object


#### Release Notes

<!-- BEGIN RELEASE NOTES -->

* polys
    * itermonomials is now a generator.
    * itermonomials now accepts (besides the list of variables) either a list of degrees (one degree for each variable) or an integer (same degree for all the variables) for either the `min_degrees` and `max_degrees` parameters. `min_degrees` can also be omitted, in which case it corresponds to a 0 (single int) or a list of 0's.

<!-- END RELEASE NOTES -->

ctsiagkalis committed:
Updated itermonomials to compute only the necessary monomials and return a generator

jksuom commented:
There is trailing whitespace that should be removed:

    File contains trailing whitespace: /home/travis/build/sympy/sympy/sympy/polys/monomials.py, line 43.

This is the first error, there may be others. You can find them by running `bin/test quality`.

asmeurer commented on sympy/polys/monomials.py:
```diff
@@ -57,37 +66,85 @@ def itermonomials(variables, max_degree, min_degree = 0):
+        if max_degree < 0 or min_degree > max_degree:
+            return set()
+        if not variables or max_degree == 0:
+            return {S(1)}
```
I think it should always return an iterator, even in the trivial cases, for consistency.

ctsiagkalis replied:
I updated the code to use yileds, but now the tests don't work because of incompatibility of `len()` function with generators.

asmeurer commented:
Please try to write instructive release notes. I updated it, but please check that what I wrote is correct and that I didn't miss anything.

asmeurer commented:
Perhaps we should make the function a generator (with `yield` statements instead of `return`).

ctsiagkalis committed:
Fixed trailing whitespace.

ctsiagkalis committed:
Updatet itermonomials so that it always returns a generator object.

ctsiagkalis committed:
Fixed more trailing whitespace.

ctsiagkalis commented:
Thanks for all your comments and suggestions.  
I updated `itermonomials` to return always a generator object (even in the trivial cases as you suggested), and also updated the release notes. 
@asmeurer 
I seem to have some issues with the tests. I tried running `bin/test quality monomials.py` or `python bin/test quality monomials py` but it always shows no error and passes the tests. Am I doing this wrong?

jksuom commented:
Changing the return type to generator has caused some failures. One of the places that have to be fixed is here https://github.com/sympy/sympy/blob/136766ae46ba07fd8ee6c2c5a039f88cd7ff1c5f/sympy/integrals/heurisch.py#L566-L569
On the two lines, `monoms` should be defined as `tuple(itermonomials(...))`. That should take care of a number of errors. You can check them locally by running `bin/test heurisch`.

ctsiagkalis committed:
Changed lines 567 and 569 to create monomials tuples
from generators

ctsiagkalis committed:
Added `if` statemnt to check for negative
`min_degrees` and `max_degrees`.
Negative `max_degrees` raises `ValueError` instead of
returning an empty Set.
Split double conditions in `if` statements to multiple `if`'s.

ctsiagkalis committed:
Updated tests for itermonomials to support generator
objects, using `next()` and `set()`.
Also, added `raises()` for new `ValueError`'s in case of negative
`max_degrees`.

ctsiagkalis committed:
Merge branch 'master' into master

ctsiagkalis committed:
Updated docstrings in monomials.py

Some old examples didn't work with the generators implementation.
Fixed them forcing list/set to match each example's result.

There was also an error in line 89 while the online tester was parsing
the file and finding EOF. I added a "\" character to hopefully escape
the whitespace.

ctsiagkalis committed:
Fixed extra character in test_monomials.py
Removed extra comma causing error in tests from line 64.

ctsiagkalis committed:
Merge branch 'master' of https://github.com/ctsiagkalis/sympy

ctsiagkalis committed:
Second update for the online tester
The "\" character didn't work, so I merged the two lines.

ctsiagkalis committed:
FIxed typo
Title underline in docstring of itermonomials was too short.

ctsiagkalis committed:
Fixed another title underline in docstring

jksuom commented on sympy/polys/monomials.py:
```diff
@@ -50,44 +59,102 @@ def itermonomials(variables, max_degree, min_degree = 0):
+        total_degree = True
+    if total_degree:
+        if min_degree > max_degree:
+            yield set()
```
I think that this should be just `return` (or `return None`). Python should raise StopIteration since there are `yield` statements elsewhere.

ctsiagkalis replied:
`>>> for i in itermonomials([x,y], 1, 2):
... print(i)
... set()`

ctsiagkalis replied:
I think it is better as it is right now, because when using the function the user can see the empty set returned:

`from sympy.abc import x, y`
`>>> for i in itermonomials([x,y], 1, 2):`
`...     print(i)`
`...`
`>>> set()`

While using only the return statement the user sees nothing returned:

`from sympy.abc import x, y`
`>>> for i in itermonomials([x,y], 1, 2):`
`...     print(i)`
`...`
`>>>`

ctsiagkalis replied:
You are actually right, the `yield` statement is redundant.
Thanks for your suggestion.

jksuom commented on sympy/polys/monomials.py:
```diff
@@ -50,44 +59,102 @@ def itermonomials(variables, max_degree, min_degree = 0):
+        if min_degree > max_degree:
+            yield set()
+        if not variables or max_degree == 0:
+            yield {S(1)}
```
Here a single item is returned (not a set) and then return added to indicate StopIteration.

    yield S(1)
    return

ctsiagkalis replied:
Again, thanks for the suggestion. It should be a bit faster like this.

jksuom replied:
This should yield the number `S(1)`, not the set `{S(1)}`. The tests below should also have `set` instead of `next`.

ctsiagkalis committed:
Small improvements in monomials.py
Replaced redundant `yield` statement with `return`.
Added return statement when returning a single element
making the generator terminating a bit faster.

jksuom commented on sympy/polys/monomials.py:
```diff
@@ -50,44 +59,103 @@ def itermonomials(variables, max_degree, min_degree = 0):
+        if any(min_degrees[i] > max_degrees[i] for i in range(n)):
+            raise ValueError('min_degrees[i] must be <= max_degrees[i] for all i')
+        power_lists = []
+        for var, min_d, max_d in zip(variables,min_degrees,max_degrees):
```
This is a minor point, but the argument list of zip would look less crowded if a space is added after the two commas.

ctsiagkalis replied:
Thanks again, I fixed that.

jksuom commented:
Can you add tests for the degree list cases as well?

ctsiagkalis committed:
Added whitespace for readability

ctsiagkalis committed:
Added extra tests mostly for degree_list case.

ctsiagkalis committed:
Fixed trailing whitespace.

jksuom commented:
The new test examples look good. One change would still be needed: `yield {S(1)}` should be `yield S(1)`, that is, it should return the number `S(1)` instead of the set `{S(1}`. The relevant examples should then work with `set(...)` instead of `next(...)`.

ctsiagkalis commented:
> 
> 
> The new test examples look good. One change would still be needed: `yield {S(1)}` should be `yield S(1)`, that is, it should return the number `S(1)` instead of the set `{S(1}`. The relevant examples should then work with `set(...)` instead of `next(...)`.

You are right. I'll change it for consistency in the tests. I will also change lines 38-40 to use `set()` instead of `next()`.

ctsiagkalis committed:
Return S(1) instead of {S(1)} and update in tests:

In `total_degree` case if the `variables` list is empty or
the `max_degree` is 0 `S(1)` should be returned instead of `{S(1)}`.
Also updated the tests to support the above change.

jksuom commented:
I think this is ready to be merged. The old total degree part could probably be implemented better as a generator but that is not necessary in this PR. Thank you for your contribution.

ctsiagkalis commented:
> 
> 
> I think this is ready to be merged. The old total degree part could probably be implemented better as a generator but that is not necessary in this PR. Thank you for your contribution.

I am glad I helped. This was my first PR and I have to admit that you helped me very much. Thank you for everything.
